langchain/src/vectorstores/azuresearch.ts

```diff
@@ -0,0 +1,562 @@
+  async addDocuments(
+    documents: Document<AzureSearchDocumentMetadata>[],
+    options?: AzureSearchAddDocumentsOptions
+  ) {
+    const texts = documents.map(({ pageContent }) => pageContent);
+
+    // Some providers like Azure OpenAI can only embed 16 documents at a time.
+    const results: string[] = [];
+    const batchSize = 16;
+
+    for (let i = 0; i < texts.length; i += batchSize) {
+      const batch = texts.slice(i, i + batchSize);
+      const docsBatch = documents.slice(i, i + batchSize);
+      const batchEmbeddings: number[][] = await this.embeddings.embedDocuments(batch);
+
+      const batchResult = await this.addVectors(
+        batchEmbeddings,
+        docsBatch,
+        options
+      );
+
+      results.push(...batchResult);
+    }
+
+    return results;
+  }
+
+  /**
+   * Upload vectors into vector store.
+   *
+   * @param vectors
+   * @param documents
+   * @param options
+   */
+  async addVectors(
+    vectors: number[][],
+    documents: Document<AzureSearchDocumentMetadata>[],
+    options?: AzureSearchAddDocumentsOptions
+  ) {
+    const keys = options?.keys ?? documents.map(() => uuid.v4());
+
+    const entities: AzureSearchDocument[] = documents.map((doc, idx) => ({
+      id: keys[idx],
+      content: doc.pageContent,
+      content_vector: vectors[idx],
+      metadata: {
+        source: doc.metadata?.source,
+        attributes: doc.metadata?.attributes ?? [],
+      }
+    }));
+
+    const chunkSize = 100;
+    for (let i = 0; i < entities.length; i += chunkSize) {
+      const chunk = entities.slice(i, i + chunkSize);
+
+      await this.client.uploadDocuments(chunk, { throwOnAnyFailure: true });
+    }
+
+    return keys;
+  }
+
+  /**
+   * Delete multiple documents by filter expression.
+   *
+   * @param filter OData filter to find documents to delete.
+   * @returns
+   */
+  async deleteMany(filter: string): Promise<IndexingResult[]> {
+    const { results } = await this.client.search("", {
+      filter,
+    });
+
+    const keys: string[] = [];
+    for await (const item of results) {
+      keys.push(item.document.id);
+    }
+
+    const { results: deleteResults } = await this.client.deleteDocuments(DEFAULT_FIELD_ID, keys);
+
+    return deleteResults;
+  }
+
+  /**
+   * Delete document by key(s).
+   *
+   * @param key
+   * @returns
+   */
+  async deleteByKey(key: string | string[]): Promise<IndexingResult[]> {
+    const { results } = await this.client.deleteDocuments(DEFAULT_FIELD_ID, Array.isArray(key) ? key : [key]);
+
+    return results;
+  }
+
+  /**
+   * Perform a similarity search using query type specified on Configuration.
+   *
+   * @param query
+   * @param k
+   * @param filter
+   */
+  async similaritySearch(
+    query: string,
+    k = 4,
+    filter: this["FilterType"] | undefined = undefined,
+  ): Promise<Document[]> {
+    const searchType = this.params.search.type;
+    let results: [Document, number][] = [];
+
+    if (searchType === "similarity") {
```

Comment:
We've tended to do this at the retriever level:

https://github.com/langchain-ai/langchainjs/blob/a084aae72d8b6ce34d4af4c499e97b22e4d6a29f/langchain/src/vectorstores/base.ts#L100

But I'm ok with putting it here since a lot of these are quite specific.